submissions/salsita/src/reducers/scrollingReducer.js

```diff
@@ -0,0 +1,145 @@
+import buildMessage from '../messageBuilder';
+import * as Effects from '../constants/effects';
+import * as Configuration from '../constants/configuration';
+import * as SithStates from '../constants/sithStates';
+import { SithRecord } from '../constants/records';
+
+// The method is responsible for enabling/disabling up/down buttons
+export const enableControlButtons = reduction => {
+  return reduction.withMutations(mutableReduction => {
+    const loadedSithsCount = mutableReduction
+      .getIn(['appState', 'siths'])
+      .filter(sith => sith.get('state') === SithStates.LOADED)
+      .count();
+
+    if (loadedSithsCount === 1) {
+      if (mutableReduction.getIn(['appState', 'siths', 0, 'state']) === SithStates.LOADED) {
+        mutableReduction.setIn(['appState', 'disabledDown'], true);
+      } else {
+        mutableReduction.setIn(['appState', 'disabledUp'], true);
+      }
+    } else if (loadedSithsCount === 0) {
+      mutableReduction.setIn(['appState', 'disabledUp'], true);
+      mutableReduction.setIn(['appState', 'disabledDown'], true);
+    } else {
+      mutableReduction.setIn(['appState', 'disabledUp'], false);
+      mutableReduction.setIn(['appState', 'disabledDown'], false);
+    }
+  });
+};
+
+export const scrolledDown = reduction => reduction.withMutations(mutableReduction => {
+  if (!mutableReduction.getIn(['appState', 'disabledDown']) && !mutableReduction.getIn(['appState', 'sithSpotted'])) {
+    // Just a flag that we are scrolling down
+    mutableReduction.setIn(['appState', 'scrollingDown'], true);
+
+    // Let's take the first N Siths which went out of the screen
+    // and find out whether they are still loading.
+    // This implementation should probably take List of Siths, not single Sith.
+    const cancelableSith = mutableReduction
+      .getIn(['appState', 'siths'])
+      .take(Configuration.PAGINATION_STEP)
+      .find(sith => sith.get('state') === SithStates.LOADING);
+
+    // If there is still a Sith which is loading, cancel its request.
+    if (cancelableSith) {
+      mutableReduction.update('effects', effects => effects.push(buildMessage(Effects.API_CANCEL_REQUEST, cancelableSith.get('correlationId'))));
```

Comment:
Don't think so, effect handler should not know anything about your domain but it works vice-versa. Actually even naming effect `API_LOAD_SITH` is wrong...